.crawler/src/crawler.rs

```
@@ -0,0 +1,274 @@
+            ..Default::default()
+        };
+
+        let pea2pea_node = Pea2PeaNode::new(Some(config)).await.unwrap();
+        let client_state = Default::default();
+        let node = Self {
+            synth_node: SynthNode::new(pea2pea_node, client_state),
+            known_network: Arc::new(KnownNetwork::default()),
+        };
+
+        node.enable_disconnect().await;
+        node.enable_handshake().await;
+        node.enable_reading().await;
+        node.enable_writing().await;
+
+        node
+    }
+
+    /// Spawns a task dedicated to broadcasting Ping messages.
+    pub fn send_pings(&self) {
+        let node = self.clone();
+        task::spawn(async move {
+            let genesis = CurrentNetwork::genesis_block();
+            let ping_msg = ClientMessage::Ping(
+                MESSAGE_VERSION,
+                MAXIMUM_FORK_DEPTH,
+                node.node_type(),
+                node.state(),
+                genesis.hash(),
+                Data::Object(genesis.header().clone()),
+            );
+
+            loop {
+                if node.node().num_connected() != 0 {
+                    info!(parent: node.node().span(), "sending out Pings");
+                    node.send_broadcast(ping_msg.clone()).unwrap();
+                }
+                tokio::time::sleep(Duration::from_secs(PING_INTERVAL_SECS)).await;
+            }
+        });
+    }
+
+    /// Spawns a task dedicated to peer maintenance.
+    pub fn update_peers(&self) {
+        let node = self.clone();
+        task::spawn(async move {
+            loop {
+                // Disconnect from peers we have just crawled.
+                for addr in node.known_network.addrs_to_disconnect() {
+                    node.node().disconnect(addr).await;
+                }
+
+                // Connect to peers we haven't crawled in a while.
+                for addr in node.known_network.addrs_to_connect() {
+                    if !node.node().is_connected(addr) {
```

this check is not needed (especially since we're ignoring errors), as `pea2pea` already does this internally.